### PR TITLE
Add options for easily setting the newsbeuter home directory.

### PR DESCRIPTION
**Pull request by tsipinakis**
This PR makes running multiple seperate instances of newsbeuter easier by adding the --homedir command line argument and NEWSBEUTERHOME environment variable to control the newsbeuter config directories.
See the commit message for more details:
> The only way to change home directories until now were the -u -c and -C
> arguments, and even with these you weren't able to change the history files.
> This commit implements the --homedir argument which switches all files to the
> directory provided. Additionally it adds support for the NEWSBEUTERHOME
> environment variable that behaves exactly as --homedir.
> 
> With this commit newsbeuter will look for its files in this order:
> 
> -u -c and -C arguments > --homedir > NEWSBEUTERHOME variable > XDG directories > $HOME/.newsbeuter

### CODE REVIEW

File: doc/manpage-newsbeuter.txt [hunk 1]
```diff
@@ -41,6 +41,9 @@ OPTIONS
 -i opmlfile, --import-from-opml=opmlfile::
        Import an OPML file
 
+-H homedir, --homedir homedir::
```
Comment (Minoru) [inline doc/manpage-newsbeuter.txt]: Please use `--homedir=homedir` format to be consistent with the rest of the manpage.

File: doc/manpage-newsbeuter.txt [hunk 2]
```diff
@@ -134,6 +137,11 @@ include::chapter-cmdline.txt[]
 '<number>'::
         Jump to the <number>th entry in the current dialog
 
+ENVIRONMENT
+----------
+
+'NEWSBEUTERHOME'::
```
Comment (Minoru) [inline doc/manpage-newsbeuter.txt]: `NEWSBEUTER_HOME` would be more readable.

File: doc/manpage-podbeuter.txt [hunk 3]
```diff
@@ -25,6 +25,9 @@ OPTIONS
 -h, --help::
         Display help
 
+-H homedir, --homedir homedir::
```
Comment (Minoru) [inline doc/manpage-podbeuter.txt]: Same as in manpage-newsbeuter: please make this `--homedir=homedir`

File: doc/manpage-podbeuter.txt [hunk 4]
```diff
@@ -54,6 +57,12 @@ CONFIGURATION COMMANDS
 include::podbeuter-cfgcmds.txt[]
 
 
+ENVIRONMENT
+----------
+
+'NEWSBEUTERHOME'::
```
Comment (Minoru) [inline doc/manpage-podbeuter.txt]: Same as in manpage-newsbeuter: please make this `NEWSBEUTER_HOME`.

File: src/controller.cpp [hunk 5]
```diff
@@ -162,25 +162,33 @@ bool controller::setup_dirs_xdg(const char *env_home, bool silent) {
...
 	const char * env_home;
-	if (!(env_home = ::getenv("HOME"))) {
-		struct passwd * spw = ::getpwuid(::getuid());
-		if (spw) {
-			env_home = spw->pw_dir;
-		} else {
-			std::cerr << _("Fatal error: couldn't determine home directory!") << std::endl;
-			std::cerr << utils::strprintf(_("Please set the HOME environment variable or add a valid user for UID %u!"), ::getuid()) << std::endl;
-			::exit(EXIT_FAILURE);
+	if(custom_home){
```
Comment (Minoru) [inline src/controller.cpp]: We're putting a space between a keyword and an opening parenthesis to distinguish them from function calls, like this:
```
if (custom_home) {
```

Please also note the space before the curly brace. Can you apply that style here and in the rest of the patchset?

File: doc/manpage-newsbeuter.txt (hunk 1 above)
Comment (Minoru) [inline doc/manpage-newsbeuter.txt]: I also think that `home-dir` might be a better fit given than we're hyphenating things like `config-file`, `url-file` and so on. What do you think?

File: src/controller.cpp [hunk 6]
```diff
@@ -252,14 +261,17 @@ void controller::run(int argc, char * argv[]) {
 
 	/* First of all, let's check for options that imply silencing of the
 	 * output: import, export, command execution and, well, quiet mode */
```
Comment (Minoru) [inline src/controller.cpp]: This comment is no longer true, can you update it?

File: src/controller.cpp [hunk 7]
```diff
@@ -252,14 +261,17 @@ void controller::run(int argc, char * argv[]) {
...
+		if(strchr("H", c) != NULL){
+			homedir = optarg;
 		}
 	}
 
-	setup_dirs(silent);
+	setup_dirs(homedir, silent);
 
 	/* Now that silencing's set up, let's rewind to the beginning of argv and
 	 * process the options */
```
Comment (Minoru) [inline src/controller.cpp]: This one is also not true since you handled not only silencing but also the homedir. Update it, please.

**Comment by Minoru** [comment]
You forgot to update usage messages.

**Comment by Minoru** [comment]
Ran the code, seems OK. Will merge once you apply my suggestions.

File: doc/manpage-newsbeuter.txt (hunk 1 above)
Comment (tsipinakis) [inline doc/manpage-newsbeuter.txt]: Agreed, I renamed it to home-dir with the latest commit.

**Comment by tsipinakis** [comment]
Made all the changes you suggested, let me know if anything else is needed.

File: doc/manpage-newsbeuter.txt [hunk 8]
```diff
@@ -41,6 +41,9 @@ OPTIONS
 -i opmlfile, --import-from-opml=opmlfile::
        Import an OPML file
 
+-H homedir, --homedir=homedir::
```
Comment (Minoru) [inline doc/manpage-newsbeuter.txt]: You've updated the code but forgot to change "homedir" to "home-dir" here.

File: doc/manpage-podbeuter.txt [hunk 9]
```diff
@@ -25,6 +25,9 @@ OPTIONS
 -h, --help::
         Display help
 
+-H homedir, --homedir=homedir::
```
Comment (Minoru) [inline doc/manpage-podbeuter.txt]: Updated the code but forgot to update the code: "home-dir", not "homedir".

**Comment by tsipinakis** [comment]
Oops, fixed.

**Comment by Minoru** [comment]
Now that I think about it more, shouldn't we steer from calling this "home directory"? Someone might become confused and thing that `newsbeuter --home-dir=/home/joe` will run Newsbeuter with Joe's config, cache file and so on. Maybe something like "config-dir" would be better? What do you think? I'm sorry for this not occurring to me earlier.

**Comment by tsipinakis** [comment]
Since XDG splits config and data directories config-dir might also mislead someone to think that its only about the config and urls and history files.

**Comment by Minoru** [comment]
Good point! I'll merge this as-is, then.

As discussed on IRC, I'm to squash your commits into one to get a prettier history, right?

**Comment by tsipinakis** [comment]
Sure, go ahead.

**Comment by Minoru** [comment]
Merged. Thank you for your work, keep it up!

**Comment by Earnestly** [comment]
Sorry, late to the party here, but why does this need to exist and what advantage does it bring over just using the `XDG_*_HOME` environments?

**Comment by Minoru** [comment]
I have to admit I didn't consider how this overlaps with XDG at the time; I only saw this features as a generalization of `-u`, `-c` and `-C` flags. One specific use-case that I had in mind is Newsbeuter development: I don't want to run the code I'm testing on my actual data, so I tend to copy it and then use `-u`, `-c` and `-C`. `--home-dir` simplifies this immensely. The fact that someone took the time to implement this persuaded me that it's a feature worth adding.

But your argument is strong, too...

I do have one thing in defence of this patchset: since environment is inherited by the subprocesses, modifying `XDG_*_HOME` might break external HTML renderers (maybe some other things too, can't day offhand).

Let's see if @tsipinakis have any more arguments.


--
Regards,
Alexander Batischev

**Comment by Earnestly** [comment]
Yeah, you do make a good point about processes newsbeuter executes; although of course you could just set their environments too, but it does get awkward I agree.

**Comment by tsipinakis** [comment]
I had the same reasoning as @Minoru to implement this. I wanted a quick and easy way to run newsbeuter on a clean environment without affecting the config and data directories of the other programs that would be used in the same shell(browser/pager etc). `newsbeuter --homedir test/` is way faster than copying 2 directories over and changing two environment variables.

**Comment by Earnestly** [comment]
Then why add a pointless `SNOWFLAKE_HOME` environment?

**Comment by Minoru** [comment]
@tsipinakis, one can alter the environment of a specific process without affecting shell's environment using `env` just like @Earnestly has shown. The following two commands are (basically) equivalent:
```
NEWSBEUTER_HOME=test ./newsbeuter
XDG_CONFIG_HOME=test XDG_DATA_HOME=test ./newsbeuter
```

The latter is a bit longer, but I don't type it all the time anyway—I just pick a line from shell's history and run it. Same goes for copying, actually: for some of my tests, I do `cp ~/.newsbeuter/cache.db test/. && ./newsbeuter -c test/cache.db` (and this is the second time in my life I had to type that in its entirety :)

There's a little difference between the two, though: XDG requires specific hierarchy, so one will have to create `newsbeuter` inside `test` directory in order for things to work. If one doesn't, Newsbeuter will happily run on the data it finds in `~/.newsbeuter`, which in my case is my live data, so I wouldn't like that.

I'm being torn between two things here:

1. On one hand, `--home-dir` is useful to me and @tsipinakis, as well as pretty much anyone who want to run more than one copy of Newsbeuter at the time (probably developers and testers).
2. On the other hand, this is a kludge aimed at a small group of people, which under other circumstances I wouldn't merge. **Furthermore**, there's pretty good alternative (XDG variables).

@tsipinakis, would you hold a grudge if I un-merge this?

**Comment by Earnestly** [comment]
> Everything was small... and my heart sinks for Linux when I see the size of
> it. [...] The manual page, which really used to be a manual page, is now a
> small volume, with a thousand options... We used to sit around in the Unix
> Room saying, 'What can we throw out? Why is there this option?' It's often
> because there is some deficiency in the basic design — you didn't really hit
> the right design point. Instead of adding an option, think about what was
> forcing you to add that option.
-- Doug MacIlroy

**Comment by Minoru** [comment]
Or, as Saint-Exupery put it, perfection is archived not when there is 
nothing more to add but when there's nothing left to remove.

It's decided, then. I'm un-merging this. Sorry, @tsipinakis.

**Pull request closed without merging**